test/VarianceTest.cc

```diff
@@ -0,0 +1,353 @@
+#undef __USE_GNU  // defined in EbThreads.h
+#endif
+#ifdef _GNU_SOURCE
+#undef _GNU_SOURCE  // defined in EbThreads.h
+#endif
+#include "aom_dsp_rtcd.h"
+#include "random.h"
+#include "util.h"
+#include "gtest/gtest.h"
+
+using svt_av1_test_tool::SVTRandom;  // to generate the random
+namespace {
+#define MAX_BLOCK_SIZE (128 * 128)
+
+static uint32_t variance_ref(const uint8_t *src, const uint8_t *ref, int width,
+                             int height, int src_stride, int ref_stride,
+                             uint32_t *sse_ptr) {
+    int64_t se = 0;
+    uint64_t sse = 0;
+    for (int y = 0; y < height; y++) {
+        for (int x = 0; x < width; x++) {
+            int diff;
+            diff = src[y * src_stride + x] - ref[y * ref_stride + x];
+            se += diff;
+            sse += diff * diff;
+        }
+    }
+    *sse_ptr = static_cast<uint32_t>(sse);
+    return static_cast<uint32_t>(sse - ((se * se) / (width * height)));
+}
+
+typedef uint32_t (*EB_TEST_MSE_NXM_TYPE)(const uint8_t *src_ptr,
+                                         int32_t src_stride,
+                                         const uint8_t *ref_ptr,
+                                         int32_t ref_stride, uint32_t *sse);
+
+typedef std::tuple<uint32_t, uint32_t, EB_TEST_MSE_NXM_TYPE> MSEParam;
+/**
+ * @brief Unit test mse functions include:
+ *  - aom_mse16x16_{c,avx2}
+ *
+ * Test strategy:
+ *  This test case use random source, ref or max source, zero ref as test
+ * pattern.
+ *
+ *
+ * Expect result:
+ *  Results come from reference functon and target function are
+ * equal.
+ *
+ * Test coverage:
+ *
+ * Test cases:
+ *
+ */
+class MseTest : public ::testing::TestWithParam<MSEParam> {
+  public:
+    MseTest() : width_(TEST_GET_PARAM(0)), height_(TEST_GET_PARAM(1)){};
+    void ref_mse() {
+        int32_t mask = (1 << 8) - 1;
+        SVTRandom *rnd = new SVTRandom(0, mask);
+        for (int i = 0; i < 10; ++i) {
+            for (int j = 0; j < MAX_BLOCK_SIZE; ++j) {
+                source_data_[j] = rnd->random();
+                ref_data_[j] = rnd->random();
+            }
+            unsigned int sse1, sse2;
+            TEST_GET_PARAM(2)(source_data_, width_, ref_data_, width_, &sse1);
+            variance_ref(source_data_,
+                         ref_data_,
+                         width_,
+                         height_,
+                         width_,
+                         width_,
+                         &sse2);
+            EXPECT_EQ(sse1, sse2);
+        }
+        delete rnd;
+    };
+    void max_mse() {
+        memset(source_data_, 255, MAX_BLOCK_SIZE);
+        memset(ref_data_, 0, MAX_BLOCK_SIZE);
+        unsigned int sse;
+        TEST_GET_PARAM(2)(source_data_, width_, ref_data_, width_, &sse);
+        const unsigned int expected = width_ * height_ * 255 * 255;
+        EXPECT_EQ(expected, sse);
+    };
+
+  protected:
+  private:
+    DECLARE_ALIGNED(32, uint8_t, source_data_[MAX_BLOCK_SIZE]);
```

Comment:
DECLARE_ALIGNED can not guarantee that the memory is aligned when the alignment is larger than 16, we should malloc a bigger buffer and aligned, you can refer aom_memalign or posix_memalign.

Reply:
modified